pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -0,0 +1,176 @@
+	sqlbase "github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/errors"
+	gogotypes "github.com/gogo/protobuf/types"
+)
+
+// Progress is streamed to the coordinator through metadata.
+var restoreDataOutputTypes = []*types.T{}
+
+type restoreDataProcessor struct {
+	flowCtx *execinfra.FlowCtx
+	spec    execinfrapb.RestoreDataSpec
+	input   execinfra.RowSource
+	output  execinfra.RowReceiver
+}
+
+var _ execinfra.Processor = &restoreDataProcessor{}
+
+// OutputTypes implements the execinfra.Processor interface.
+func (rd *restoreDataProcessor) OutputTypes() []*types.T {
+	return restoreDataOutputTypes
+}
+
+func newRestoreDataProcessor(
+	flowCtx *execinfra.FlowCtx,
+	_ int32,
+	spec execinfrapb.RestoreDataSpec,
+	input execinfra.RowSource,
+	output execinfra.RowReceiver,
+) (execinfra.Processor, error) {
+	rd := &restoreDataProcessor{
+		flowCtx: flowCtx,
+		input:   input,
+		spec:    spec,
+		output:  output,
+	}
+	return rd, nil
+}
+
+// Run implements the execinfra.Processor interface.
+func (rd *restoreDataProcessor) Run(ctx context.Context) {
+	ctx, span := tracing.ChildSpan(ctx, "restoreDataProcessor")
+	defer tracing.FinishSpan(span)
+	defer rd.output.ProducerDone()
+
+	progCh := make(chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress)
+
+	// We don't have to worry about this go routine leaking because next we loop over progCh
+	// which is closed only after the goroutine returns.
+	var err error
+	go func() {
+		defer close(progCh)
+		err = runRestoreData(ctx, rd.flowCtx, &rd.spec, rd.input, progCh)
+	}()
+
+	for prog := range progCh {
+		// Take a copy so that we can send the progress address to the output processor.
+		p := prog
+		rd.output.Push(nil, &execinfrapb.ProducerMetadata{BulkProcessorProgress: &p})
+	}
+
+	if err != nil {
+		rd.output.Push(nil, &execinfrapb.ProducerMetadata{Err: err})
+		return
+	}
+}
+
+func runRestoreData(
+	ctx context.Context,
+	flowCtx *execinfra.FlowCtx,
+	spec *execinfrapb.RestoreDataSpec,
+	input execinfra.RowSource,
+	progCh chan execinfrapb.RemoteProducerMetadata_BulkProcessorProgress,
+) error {
+	input.Start(ctx)
+	kr, err := storageccl.MakeKeyRewriterFromRekeys(spec.Rekeys)
+	if err != nil {
+		return err
+	}
+
+	for {
+		// We read rows from the SplitAndScatter processor. We expect each row to
+		// contain 2 columns. The first is used to route the row to this processor,
+		// and the second contains the RestoreSpanEntry that we're interested in.
+		row, meta := input.Next()
+		if meta != nil {
+			return errors.Newf("unexpected metadata %+v", meta)
+		}
+		if row == nil {
+			// Done.
+			break
+		}
+
+		if len(row) != 2 {
+			return errors.New("expected input rows to have exactly 2 columns")
+		}
+		alloc := &sqlbase.DatumAlloc{}
```

Comment:
nit: could pull this out of the loop so it gets reused as intended.

Reply:
Done.